### PR TITLE
Nuisance Regression Parameters

### PR DESCRIPTION
**Pull request by samlarson**
WIP PR 
* Added table of contents for repo readme
* Added getopts flag for nuisance regression

### CODE REVIEW

File: PreprocessingScript_step2.sh [hunk 1]
```diff
@@ -14,6 +14,35 @@ sm_sigma="2.1233226" # spatial smoothing sigma
...
+  printf "        b: 10 Principle Components (non-brain tissues)\n"
+  printf "        c: 6 Motion Regressors (based on motion correction)\n"
+  printf "        d: 6 Motion Derivative Regressors (temporal derivatives of c)\n"
+  printf "        e: CSF & WMCSF Signals\n"
+  printf "    [Example]\n"
+  printf "        ./preproc-script-2.sh -n abc\n\n"
+  printf "    -h    Help (displays these usage details)\n"
+  printf "    [Example]\n"
+  printf "        ./preproc-script-2.sh -h\n\n"
+}
```
Comment (imnanxu) [inline PreprocessingScript_step2.sh]: 1. change the list of options as the following and perhaps make the options more intuitive:
  printf "Options:\n"
  printf "    -nuis    Nuisance Regression Parameters (combinations supported)\n"
  printf "    [Values]\n"
  printf "        trends: 3 Detrends (constant/linear/quadratic trends)\n"
  printf "        mot: 6 Motion Regressors (based on motion correction)\n"
  printf "        motder: 6 Motion Derivative Regressors (temporal derivatives of c)\n"
  printf "        csf: CSF Signals\n"
  printf "        wmcsf: WMCSF Signals only valid for rat brains\n"
  printf "        10pca: 10 Principle Components (non-brain tissues)\n"
  printf "        spca: Selected Principle Components (non-brain tissues) will add more here, but let's leave it for now\n"
  printf "    [Note:] By default, nuisance regressions with global signals and without will be generated.\n"
  printf "    [Example]\n"
 printf "        ./preproc-script-2.sh --nuis=trends, mot, spca, csf\n\n"

Q: Can we let the user freely choose the order of these nuisances?

2. we will also add other parameters, e.g.,
"./preproc-script-2.sh --model=rat --datafolder=data1, data2 --fil=0.01, 0.02 --smfwhm=5  --nuis=trends, mot, spca, csf"

File: PreprocessingScript_step2.sh [hunk 2]
```diff
@@ -73,15 +102,31 @@ do
...
+    printf "\n"
+  fi
+  if [[ $nuisance == *"d"* ]]; then
+    # Calculate motion derivative regressors (AFNI)---works for small motion defects
+	  1d_tool.py -overwrite -infile ./"$workingdir"/EPI_mc.par -derivative -write ./"$workingdir"/motionEPI.deriv.par
+  fi
+  if [[ $nuisance == *"e"* ]]; then
+    # extract CSF/WMCSF/Global signals---noise
+	  fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/csfEPI.txt -m ./"$workingdir"/EPI_n4_csf_mask.nii.gz
+	  fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/gsEPI.txt -m ./"$workingdir"/EPI_n4_mask.nii.gz
```
Comment (imnanxu) [inline PreprocessingScript_step2.sh]: This is a default option, so remove it from here and leave Ln83 as it was.

File: PreprocessingScript_step2.sh [hunk 3]
```diff
@@ -73,15 +102,31 @@ do
...
 	# ##-------------Nuisance regressors---------------------------
 	echo "====================$workingdir: Nuisance regression: motions, motion devs, PCA noise, trends, gs/wmcsf/csf===================="
-	# Calculate motion derivative regressors (AFNI)---works for small motion defects
-	1d_tool.py -overwrite -infile ./"$workingdir"/EPI_mc.par -derivative -write ./"$workingdir"/motionEPI.deriv.par
-	# get constant, linear, quad trends
-	NR=$(3dinfo -nv ./"$workingdir"/EPI_topup.nii.gz);
-	1dBport -band 0 0 -nodata ${NR} -quad > ./"$workingdir"/quad_regressionEPI.txt
-	# extract CSF/WMCSF/Global signals---noise
-	fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/csfEPI.txt -m ./"$workingdir"/EPI_n4_csf_mask.nii.gz	
-	fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/gsEPI.txt -m ./"$workingdir"/EPI_n4_mask.nii.gz
```
Comment (imnanxu) [inline PreprocessingScript_step2.sh:83]: This is the default option, so we need this no matter what was specified in the nuisance parameters.

File: PreprocessingScript_step2.sh [hunk 4]
```diff
@@ -73,15 +102,31 @@ do
...
+  fi
+  if [[ $nuisance == *"e"* ]]; then
+    # extract CSF/WMCSF/Global signals---noise
+	  fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/csfEPI.txt -m ./"$workingdir"/EPI_n4_csf_mask.nii.gz
+	  fslmeants -i ./"$workingdir"/EPI_topup.nii.gz -o ./"$workingdir"/gsEPI.txt -m ./"$workingdir"/EPI_n4_mask.nii.gz
+  fi
+
+
+
+	# TODO: paste operation arguments need to be conditional based on above criteria
```
Comment (imnanxu) [inline PreprocessingScript_step2.sh]: Good!

File: README.md [hunk 5]
```diff
@@ -3,11 +3,38 @@ This is a generally applicable fMRI preprocessing toolbox for the whole brain of
 
 If you find this toolbox useful for your work, please cite as: Nan Xu, Leo Zhang, Zengmin Li, Wen-Ju Pan, Kai-Hsiang Chuang, Shella Keilholz, 2021. Rodent whole-brain fMRI data preprocessing toolbox. https://github.com/GT-EmoryMINDlab/rodent-whole-brain-preprocessing-recipe
 
-## I. Prerequisite software
+# Table of Contents
```
Comment (imnanxu) [inline README.md:6]: I like the new TOC!

File: PreprocessingScript_step2.sh (hunk 1 above)
Comment (samlarson) [reply PreprocessingScript_step2.sh]: The builtin getopts doesn't support long flags (flags can only be one character long), but it does allow the user to freely choose the order. We can workaround this by parsing long flags in a loop and passing them to getopts in bash (it's a bit ugly), or we can discuss using a different tool to handle CLI options.

Comment (justbennet) [reply PreprocessingScript_step2.sh]: You might wish to look at `getopt` which does have long options.  It should be available on most Linux.  There is a good example on this page.

https://www.tutorialspoint.com/unix_commands/getopt.htm

Comment (samlarson) [reply PreprocessingScript_step2.sh]: @nxu25 I added a commit to make long flags work with getopts (comma delimited). I also changed the name of the common nuisance regressor files from 'gsEPI_nuisance_design.txt' and 'csfEPI_nuisance_design.txt' to a singular 'nuisance_design.txt', let me know if that's acceptable or if you want certain options to write to different csf/gs files. I have not made changes to those file references (below the nuisance regressor calculations).

**Comment by samlarson** [comment]
Closing this PR, all commits from this PR were merged into `main` via #2.

**Pull request closed without merging**